### ISSUE TITLE
Command structure is annoyingly unmovable

### ISSUE THREAD

**Issue by schveiguy** [mysql-d/mysql-native#75]
The struct `Command` in mysql-native needs to stay put as you are receiving results. This makes it difficult to abstract mysql queries into, for instance, a range of objects that materialize one at a time as rows are received. This is because the `ResultSequence` struct stores a pointer to the `Command` struct, and uses it during iteration. Having pointers to structs means they have to stay put, but structs easily move around in D (resulting in easy memory errors).

Is there a way `Command` can be turned into a ref-counted struct placed on the C heap? Or perhaps the elements of `Command` that pertain to iteration can be moved to `ResultSequence`?

**Comment by schveiguy** [comment]
Wow, actually, the command's parameters also need to stay put, so this is really anchored into place. Abstraction is terribly difficult with this model.

**Comment by Abscissa** [comment]
Can you give a common example of when a Command struct would need to move?

**Comment by schveiguy** [comment]
So for example, I want to generate a sequence of rows given a connection amd the SQL. I can't just pass in sql, I have to pass in a Command. One line foreach such as this is not possible:

```D
foreach(r; someCommandSetupThing("SELECT * FROM theTable"))
```

Nope, because a Command struct must be generated from the SQL, and THAT must not move, since the ResultSequence contains a reference to that Command struct. This makes abstracting the database much more difficult.

However, once I've executed the command, there's no reason I need the command, or all the pieces of it, right? I just need the connection to fetch more data.

Alternatively, the command can simply be copied into the ResultSequence instead of referenced. But that's not the whole problem, as I said in my second comment. Not only can the command not move, but the parameters need to stay put as well. Pipelining statements a la ranges is nigh impossible.

**Comment by Abscissa** [comment]
These issues should be fixed with the [new design](http://forum.dlang.org/thread/o6ma0r$20n6$1@digitalmars.com) as long as you don't use the deprecated (and no longer necessary) Command struct.

In the new design, neither ResultSet nor ResultSequence (now named ResultRange) hold any pointers, or go anywhere near Command. And with the new prepared statement struct, Prepared, parameters are stored by value, not by reference/ptr. So there *shouldn't* be any issue left, but please go ahead and double-check.

**Comment by schveiguy** [comment]
I'm coincidentally about to get back into a project that uses mysql-native. I'll take a look at the new API and see how it stands with my current code. Thanks!

**Comment by Abscissa** [comment]
Fixed in v1.0.0
